File: backend/src/main/java/com/bcodesphere/pilot/compartido/Dinero.java
package com.bcodesphere.pilot.compartido;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.regex.Pattern;

/**
 * Monto en USD con escala fija de 2 decimales (ADR-022) y redondeo HALF_UP (CLAUDE.md 1.1.2).
 * Admite negativos porque también representa diferencias y saldos con signo (esquema {@code MontoConSigno});
 * exigir "no negativo" es responsabilidad de quien lo necesite (esquema {@code Monto}).
 * Nunca expone {@code double} ni {@code float} (CLAUDE.md 1.2.2).
 *
 * @param valor cantidad ya normalizada a 2 decimales
 */
public record Dinero(BigDecimal valor) implements Comparable<Dinero> {

    /** Cero con escala 2. */
    public static final Dinero CERO = new Dinero(BigDecimal.ZERO);

    /** Patrón de {@code MontoConSigno} del contrato OpenAPI: hasta 17 enteros y 2 decimales. */
    private static final Pattern PATRON = Pattern.compile("^-?\\d{1,17}(\\.\\d{1,2})?$");

    /** Escala contable única (ADR-022). */
    private static final int ESCALA = 2;

    /**
     * Normaliza a escala 2 sin redondear en silencio: un valor con más decimales significativos se rechaza,
     * para que nadie pierda centavos sin decidirlo (usar {@link #redondear(BigDecimal)} para cálculos).
     *
     * @throws IllegalArgumentException si el valor es nulo o tiene más de 2 decimales significativos
     */
    public Dinero {
        // 1. El valor es obligatorio
        if (valor == null) {
            throw new IllegalArgumentException("El monto no puede ser nulo");
        }
        // 2. Se elimina el ruido de ceros a la derecha (100.000 -> 100) para medir los decimales reales
        BigDecimal limpio = valor.stripTrailingZeros();
        if (limpio.scale() > ESCALA) {
            throw new IllegalArgumentException("El monto no admite más de 2 decimales");
        }
        // 3. Escala fija: 100.0 y 100.00 quedan idénticos, así equals compara por valor
        valor = valor.setScale(ESCALA, RoundingMode.UNNECESSARY);
    }

    /**
     * Interpreta una cadena decimal del contrato ({@code ^-?\d{1,17}(\.\d{1,2})?$}).
     *
     * @param texto monto como cadena, por ejemplo {@code "123.45"} o {@code "-10.5"}
     * @return el monto normalizado a 2 decimales
     * @throws IllegalArgumentException si no cumple el patrón (notación científica, más de 2 decimales, etc.)
     */
    public static Dinero de(String texto) {
        // 1. Solo el patrón exacto del contrato; nada de notación científica ni espacios
        if (texto == null || !PATRON.matcher(texto).matches()) {
            throw new IllegalArgumentException("Monto inválido: se esperaba una cadena decimal con máximo 2 decimales");
        }
        return new Dinero(new BigDecimal(texto));
    }

    /**
     * Redondea el resultado de un cálculo (p. ej. IVA) a 2 decimales con HALF_UP.
     *
     * @param resultado valor calculado con cualquier escala
     * @return monto con escala 2
     */
    public static Dinero redondear(BigDecimal resultado) {
        if (resultado == null) {
            throw new IllegalArgumentException("El monto no puede ser nulo");
        }
        return new Dinero(resultado.setScale(ESCALA, RoundingMode.HALF_UP));
    }

    /** Suma exacta de dos montos. */
    public Dinero sumar(Dinero otro) {
        return new Dinero(valor.add(otro.valor));
    }

    /** Resta exacta {@code this - otro}; puede dar negativo. */
    public Dinero restar(Dinero otro) {
        return new Dinero(valor.subtract(otro.valor));
    }

    /** Indica si el monto es exactamente cero. */
    public boolean esCero() {
        return valor.signum() == 0;
    }

    /** Indica si el monto es mayor que cero. */
    public boolean esPositivo() {
        return valor.signum() > 0;
    }

    /** Compara por valor con otro monto; negativo, cero o positivo según {@code this} sea menor, igual o mayor. */
    public int comparar(Dinero otro) {
        return valor.compareTo(otro.valor);
    }

    @Override
    public int compareTo(Dinero otro) {
        return comparar(otro);
    }

    /** Representación en cadena decimal con 2 decimales, la misma que viaja en la API (ADR-013). */
    @Override
    public String toString() {
        return valor.toPlainString();
    }
}

File: frontend/src/compartido/dinero/dinero.ts
import Decimal from 'decimal.js';

/**
 * Constructor de Decimal propio del módulo. El valor por defecto de decimal.js (precision = 20 dígitos
 * significativos) redondea en silencio: un monto válido llega a 19 dígitos y sus sumas pueden pasarse.
 * Se usa un clon con precision 50 (sin tocar `Decimal.set`, para no afectar otro código global)
 * y redondeo HALF_UP (CLAUDE.md §1.1.2).
 */
const DecimalDinero = Decimal.clone({ precision: 50, rounding: Decimal.ROUND_HALF_UP });

/** Patrón de `Monto` del contrato OpenAPI: no negativo, hasta 17 enteros y máximo 2 decimales (ADR-013). */
const PATRON_MONTO = /^\d{1,17}(\.\d{1,2})?$/;

/**
 * Convierte una cadena a Decimal validando el patrón del contrato.
 * @param monto cadena decimal (p. ej. "123.45")
 * @throws Error si la cadena no cumple el patrón (p. ej. "1.005", "-1", "abc")
 */
function aDecimal(monto: string): Decimal {
  // 1. Se rechaza todo lo que el backend también rechazaría
  if (!PATRON_MONTO.test(monto)) throw new Error(`Monto inválido: "${monto}"`);
  return new DecimalDinero(monto);
}

/**
 * Suma montos con aritmética decimal exacta.
 * @param montos cadenas decimales válidas
 * @returns la suma con exactamente 2 decimales (p. ej. ["0.10","0.20"] → "0.30")
 */
export function sumarMontos(montos: string[]): string {
  // 1. Suma con decimal.js: nunca con number, para evitar errores de coma flotante
  const total = montos.reduce((acumulado, m) => acumulado.plus(aDecimal(m)), new DecimalDinero(0));
  return total.toFixed(2, Decimal.ROUND_HALF_UP);
}

/**
 * Compara dos montos por valor, ignorando la escala ("100.0" = "100.00").
 * @param a primer monto
 * @param b segundo monto
 */
export function sonIguales(a: string, b: string): boolean {
  return aDecimal(a).equals(aDecimal(b));
}

/**
 * Formatea un monto como moneda `$1,234.56` sin pasar por number (no pierde precisión con 17 dígitos).
 * @param monto cadena decimal válida
 */
export function formatearMoneda(monto: string): string {
  // 1. Fija 2 decimales con redondeo HALF_UP (CLAUDE.md §1.1.2)
  const [enteros = '0', decimales = '00'] = aDecimal(monto).toFixed(2, Decimal.ROUND_HALF_UP).split('.');
  // 2. Agrupa los miles a mano (el resultado no depende de los datos de Intl)
  const agrupados = enteros.replace(/\B(?=(\d{3})+(?!\d))/g, ',');
  return `$${agrupados}.${decimales}`;
}
